beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -0,0 +1,163 @@
+package sync
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/runutil"
+	"github.com/prysmaticlabs/prysm/shared/traceutil"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+	"golang.org/x/exp/rand"
+)
+
+// This defines how often a node cleans up and processes pending attestations in the queue.
+var processPendingAttsPeriod = time.Duration(params.BeaconConfig().SecondsPerSlot/2) * time.Second
+
+// This processes pending attestation queues on every `processPendingAttsPeriod`.
+func (s *Service) processPendingAttsQueue() {
+	ctx := context.Background()
+	runutil.RunEvery(s.ctx, processPendingAttsPeriod, func() {
+		s.processPendingAtts(ctx)
+	})
+}
+
+// This defines how pending attestations are processed. It contains features:
+// 1. Clean up invalid pending attestations from the queue.
+// 2. Check if pending attestations can be processed when the block has arrived.
+// 3. Request block from a random peer if unable to proceed step 2.
+func (s *Service) processPendingAtts(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "processPendingAtts")
+	defer span.End()
+
+	pids := s.p2p.Peers().Connected()
+
+	// Before a node processes pending attestations queue, it verifies
+	// the attestations in the queue are still valid. Attestations will
+	// be deleted from the queue if invalid (ie. getting staled from falling too many slots behind).
+	s.validatePendingAtts(ctx, s.chain.CurrentSlot())
+
+	for bRoot, attestations := range s.blkRootToPendingAtts {
+		// Has the pending attestation's missing block arrived yet?
+		if s.db.HasBlock(ctx, bRoot) {
+			numberOfBlocksRecoveredFromAtt.Inc()
+			for _, att := range attestations {
+				// The pending attestations can arrive in both aggregated and unaggregated forms,
```

Comment:
this whole following chunk has multiple conditionals and seems to be doing a lot. Can it be abstracted into smaller methods ?

Reply:
It's actually not bad, it's just comments that seem long. I prefer to keep it as it is for clear commentary.  Each if else is just 3-4 lines long